BAMF - NNUnet Liver MR

rahul99 opened the pull request.

Commit:
resources for bamf-nnunet-mr-liver model

Commit:
update config values and zenodo url

Commit:
refactor dockerfile

Commit:
fix modality val in runner

vanossj committed:
add some in progress metadata

vanossj committed:
remove comments

vanossj committed:
add preprint info

vanossj commented:
Hi @LennyN95, We added the preprint info into the meta.json. This PR should be good to go. Let us know if there is are any updated to the code or meta.json we need to make to keep current with the latest mhub api/requirements

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,35 @@
+    source_segs: 'nifti:mod=seg'
+    model_name: 'Bamf NNUnet MR Liver'
+    skip_empty_slices: True
+    json_config_path: '/app/models/bamf_nnunet_mr_liver/config/meta.json'
```
For dicom conversion there are two options
- provide a json definition file 
- use our MHub segmentation database (SegDB)

The former use is for all ROI that are not included in our DB (and we will modify the implementation once we added them). As we have the liver in our SegDB (LIVER) and you already linked the ROI correctly (see line 25, the NNUnet Runner will tag the foreground segmentation with LIVER), we can delete this line `json_config_path: '/app/models/bamf_nnunet_mr_liver/config/meta.json'` and also the JSON definition file under `models/bamf_nnunet_mr_liver/config/meta.json`.

LennyN95 commented:
Hi @vanossj, thank you for the update! 

I reviewed the model and found the following open issues:
- You define a `BamfProcessorRunner` under `scripts/BamfProcessor.py`, but it's not used in the config.yml. Is this a required post-processing step?
- [x] The `DsegConverter` should use the ROI instead of a custom JSON as the ROI is available in our SegDB ([read here](https://github.com/MHubAI/documentation/blob/main/documentation/mhubio/mhubio_modules.md#dsegconverter)). You can remove the `DsegConverter.json_config_path` argument.
- [x] The `config/meta.json` can be removed.
- [x] Remove `/scripts` as all MHub models are required to provide a default.yml entry point only (we support scripting within our framework but do not allow it in our models repository for standardization purposes)
- [x] Remove all `COPY`and `ADD` instructions from the Dockerfile. We do not allow those.
- [x] The Dockerfile Entrypoint must be `ENTRYPOINT ["python3", "-m", "mhubio.run"]`. The `CMD -["--config",  "/app/path/to/default.yml"]` then must point to your default config.

In addition, we will require a public example to verify the model works correctly. 
- Can you list one (or multiple) suitable IDC collections that should be supported by the model?
- Can you run the MHub implementation of the model on 2-3 example cases, and share the link to the examples with us as well as the generated output (untouched)?

I'm currently working on an automation pipeline. In the near future, sample data will likely be referenced in the `meta.json` ;) 

Thank you for your contribution!

Commit:
update BamfProcessor to inherit from mhub module

LennyN95 commented on models/bamf_nnunet_mr_liver/config/meta.json:
File can be removed.

LennyN95 commented on models/bamf_nnunet_mr_liver/meta.json:
```diff
@@ -0,0 +1,135 @@
+            "external": true
+        }
+    },
+    "details": {
```
`cite` is missing.

LennyN95 replied:
Apparently there are different rules depending on the country, here we use license as the verb and licence as the noun ^^

Knowing this, using license as keyword likely would've been the easier choice, but for now could you just change `license` to `licence` so the checks pass :)?

LennyN95 commented on models/bamf_nnunet_mr_liver/scripts/BamfProcessor.py:
Remove file.

LennyN95 commented on models/bamf_nnunet_mr_liver/scripts/run.py:
remove file

LennyN95 commented on models/bamf_nnunet_mr_liver/scripts/slicer_run.py:
Remove file.

LennyN95 commented on models/bamf_nnunet_mr_liver/utils/BamfProcessorRunner.py:
```diff
@@ -0,0 +1,84 @@
+import shutil
+
+
+@IO.ConfigInput('in_data', 'nifti:mod=ct|mr', the="input data to run BamfProcessor on")
```
See the comments I added regarding BamfProcessor in [PR50](https://github.com/MHubAI/models/pull/50).

Commit:
address changes suggested by mhub dev

Commit:
update meta

Commit:
add slicethickness

Commit:
- In /summary/model/training
    Supervised to supervised
- In /summary/model/cmpapproach
    "#D, Emsemble" to "3D"
- /details/licence to /details/license

Commit:
fix: buildutils

Commit:
update BamfProcessor

Commit:
Update Bamf processor

Commit:
explicitly convert mr dicoms to nifti

Commit:
explicitly set target as mr for DsegConverter

fedorov commented:
@rahul99 for testing the model, please provide the following for a sample MR series that is known to produce acceptable results with this model:

*`SeriesInstanceUID`
* `aws_url` for the bucket folder containing the series files
* version of IDC data that you used for getting the above

If you pick one of the cases that are already accompanied by a segmentation produced by this model in IDC (they have just been added yesterday in IDC v17), you can also include the items above for the SEG series.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+    import_dir: sorted_data
+    sort_data: true
+    meta: 
+      mod: 'mr'
```
I suggest using the `%Modality` placeholder here, our DicomImporter module then will infer the Modality from the DICOM headers automatically, preventing e.g. CT scans from being processed.

Suggested change:  `mod: '%Modality'`.

**Note, that quotes are required as the string starts with a special character `%`.**

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+
+  NiftiConverter:
+    in_datas: dicom:mod=mr
+    engine: dcm2niix
```
Is there any particular reason to use the `dcm2niix` engine over `plastimatch`?

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+    
+  NNUnetRunner:
+    in_data: nifti:mod=mr
+    nnunet_task: 'Task822_Liver_AMD'
```
I suggest to remove quotes for readability. 

Suggested change: `nnunet_task: Task822_Liver_AMD`

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+  NNUnetRunner:
+    in_data: nifti:mod=mr
+    nnunet_task: 'Task822_Liver_AMD'
+    nnunet_model: '3d_fullres'
```
I suggest to remove quotes for readability.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+    nnunet_model: '3d_fullres'
+    roi: LIVER
+
+  BamfProcessorRunner:
```
Line can be removed if no configuration is specified (will still be executed, as it's listed in the `execute` chain).

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+  BamfProcessorRunner:
+
+  DsegConverter:
+    source_segs: 'nifti:mod=seg'
```
I suggest to remove quotes for readability (quotes are only required if the line starts with special characters. Wheather quotes are used or not should be consistent. I suggest to not use them when not required.)

LennyN95 commented on models/bamf_nnunet_mr_liver/config/default.yml:
```diff
@@ -0,0 +1,42 @@
+  DsegConverter:
+    source_segs: 'nifti:mod=seg'
+    target_dicom: dicom:mod=mr
+    model_name: 'Bamf NNUnet MR Liver'
```
I suggest to remove quotes for readability.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/slicer.yml:
```diff
@@ -0,0 +1,34 @@
+general:
```
Please provide testing for this workflow when submitting the test results.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/slicer.yml:
```diff
@@ -0,0 +1,34 @@
+
+modules:
+  NrrdImporter:
+    input_dir: 'input_data'
```
I suggest to remove quotes for readability.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/slicer.yml:
```diff
@@ -0,0 +1,34 @@
+modules:
+  NrrdImporter:
+    input_dir: 'input_data'
+    input_file_name: 'image.nrrd'
```
I suggest to remove quotes for readability.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/slicer.yml:
```diff
@@ -0,0 +1,34 @@
+    - nifti:mod=seg-->[basename]
+
+  NNUnetRunner:
+    nnunet_task: 'Task882_MR_Liver'
```
I suggest to remove quotes for readability.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/slicer.yml:
```diff
@@ -0,0 +1,34 @@
+
+  NNUnetRunner:
+    nnunet_task: 'Task882_MR_Liver'
+    nnunet_model: '3d_fullres'
```
I suggest to remove quotes for readability.

LennyN95 commented on models/bamf_nnunet_mr_liver/config/slicer.yml:
```diff
@@ -0,0 +1,34 @@
+    nnunet_model: '3d_fullres'
+    roi: LIVER
+
+  BamfProcessorRunner:
```
Line can be removed.

LennyN95 commented on models/bamf_nnunet_mr_liver/scripts/BamfProcessor.py:
```diff
@@ -0,0 +1,73 @@
+"""
```
Remove file (remove the entire scripts folder).

LennyN95 commented on models/bamf_nnunet_mr_liver/scripts/run.py:
```diff
@@ -0,0 +1,56 @@
+"""
```
Remove file (remove the entire scripts folder).

LennyN95 commented on models/bamf_nnunet_mr_liver/scripts/slicer_run.py:
```diff
@@ -0,0 +1,37 @@
+"""
```
Remove file (remove the entire scripts folder).

LennyN95 commented on models/bamf_nnunet_mr_liver/utils/BamfProcessorRunner.py:
```diff
@@ -0,0 +1,66 @@
+import SimpleITK as sitk
+from skimage import measure, filters
+import numpy as np
+import shutil
```
Please omit all imports not used.

LennyN95 commented on models/bamf_nnunet_mr_liver/utils/BamfProcessorRunner.py:
```diff
@@ -0,0 +1,66 @@
+    @IO.Input('in_data', 'nifti:mod=ct|mr', the='input data to run nnunet on')
+    @IO.Output('out_data', 'bamf_processed.nrrd', 'nrrd:mod=seg:processor=bamf', data='in_data', the="keep the two largest connected components of the segmentation and remove all other ones")
+    def task(self, instance: Instance, in_data: InstanceData, out_data: InstanceData) -> None:
+       # Log bamf runner info
```
The comment needs one more space to match the indent right :D

LennyN95 replied:
I'd add an empty line above for better readability, but that's a personal preference ;)

LennyN95 commented on models/bamf_nnunet_mr_liver/utils/BamfProcessorRunner.py:
```diff
@@ -0,0 +1,66 @@
+    @IO.Output('out_data', 'bamf_processed.nrrd', 'nrrd:mod=seg:processor=bamf', data='in_data', the="keep the two largest connected components of the segmentation and remove all other ones")
+    def task(self, instance: Instance, in_data: InstanceData, out_data: InstanceData) -> None:
+       # Log bamf runner info
+        self.v("Running BamfProcessor on....")
```
As this file is used in all four models (and potential future models I suppose), I suggest to use the `self.log` function instead of the legacy `self.v`. It works pretty much the same but gives some additional control over the log-level. Documentation can be found here: https://github.com/MHubAI/documentation/blob/main/documentation/mhubio/how_to_write_an_mhubio_module.md#log-messages

LennyN95 replied:
Simply replacing `self.v` by `self.log` is already sufficient here!

LennyN95 commented on models/bamf_nnunet_mr_liver/utils/BamfProcessorRunner.py:
```diff
@@ -0,0 +1,66 @@
+"""
```
_**INFO**: Please copy all changes over to the other PR's where you use your custom `BamfProcessorRunner` Module._

Commit:
refactor module

Commit:
refactor module

rahul99 commented:
/test

```yaml
sample:
  idc_version: 17.0
  data:
  - SeriesInstanceUID: 1.3.6.1.4.1.14519.5.2.1.3344.4008.993129740958163005437786531631
    aws_url: s3://idc-open-data/124ed9dd-2085-4d8f-99aa-76b9430b2b95/*
    path: dicom

reference:
  url: https://www.dropbox.com/scl/fi/xbt99t9618aii0hv1jcwx/bamf_nnunet_mr_liver.zip?rlkey=ai8jrin43ce43wxsf4iresqv4&dl=0
```

---

<details><summary>Test Results (24.03.28_17.30.31_kb1YE6N1mX)</summary>

```yaml
id: 301ae31d-56e0-43c8-b759-5700478cbe52
date: '2024-03-28 17:59:25'
checked_files:
- file: bamf_nnunet_mr_liver.seg.dcm
  path: /app/test/src/1.3.6.1.4.1.14519.5.2.1.3344.4008.993129740958163005437786531631/bamf_nnunet_mr_liver.seg.dcm
  checks:
  - checker: DicomsegContentCheck
    notes:
    - label: Segment Count
      description: The number of segments identified in the inspected dicomseg file.
      info: 1
summary:
  files_missing: 0
  files_extra: 0
  checks:
    DicomsegContentCheck:
      files: 1
conclusion: true
```

</details>

rahul99 commented:
/test

```
sample:
  idc_version: 17.0
  data:
    - SeriesInstanceUID: 1.3.6.1.4.1.14519.5.2.1.3344.4008.993129740958163005437786531631
      aws: s3://idc-open-data/124ed9dd-2085-4d8f-99aa-76b9430b2b95/*
      path: dicom

reference:
  url: https://www.dropbox.com/scl/fi/xbt99t9618aii0hv1jcwx/bamf_nnunet_mr_liver.zip?rlkey=ai8jrin43ce43wxsf4iresqv4&dl=0
```

LennyN95 commented:
Please just edit the comment if there was a mistake and it's not ment to be a new gesture run as otherwise transparency suffers.

fedorov commented:
IMHO, new test entry on modification is a lot more transparent.

Further, (and fortunately!) it is not always possible to edit comments of other users.

But I am not the contributor, so you guys proceed as it works for you.

LennyN95 commented:
I agree with all modifications @fedorov, especially if a test has been performed already. However, for correction of mistakes, it is too much to have 2,3,4.. of these information-dense blocks because you have to find the one used in testing then carefully. If a different user proposes a different test, that's fine. I just meant small changes like a wrong key or bad formatting. So, the rule of thumb could be that you should edit your test whenever the "INVALID TEST REQUEST" label is assigned. Otherwise, create a new one.

LennyN95 commented on models/bamf_nnunet_mr_liver/meta.json:
```diff
@@ -0,0 +1,137 @@
+                "description": "The abdominal T1 weighted  MRI scan of a patient.",
+                "format": "DICOM",
+                "modality": "MR",
+                "bodypartexamined": "LIVER",
```
Please change to "Liver"

fedorov replied:
Just curious, what is the reason for deviating from the conventions in the standard regarding capitalization?

It is difficult enough to remember one set of conventions, and now one has to remember two... :-(

LennyN95 replied:
Oh, I thought it's upper-case first letter (that's how we had it in all our first models). Did we confuse this?

LennyN95 replied:
Uff, yeah we did ([list of values for body part examined](https://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_L.html#chapter_L) for reference).

LennyN95 replied:
Please ignore @rahul99. I'll fix this server-side. Sorry for the confusion and thank you Andrey for noticing this!

rahul99 replied:
@LennyN95 , I followed this advice on `bamf_nnunet_ct_kidney` PR and updated "KIDNEY" to "Kidney". Should I revert it?

LennyN95 replied:
No worries @rahul99, I updated the backend such that it is case-insensitive. I might patch all metadata at a later point to the correct all-uppercase spelling but for now it's ok ;)

LennyN95 commented on models/bamf_nnunet_mr_liver/meta.json:
```diff
@@ -0,0 +1,137 @@
+        },
+        "analyses": {
+            "title": "Quantitative Analyses",
+            "text": "The model's performance was assessed using the Dice Coefficient and Normalized Surface Distance (NSD) with tolerance 7mm, as specified in the CT Liver segmentation task in the Medical Segmentation Decathlon challenge. The model was used to segment cases from the IDC collection TCGA-LIHC [1]. Nine of those cases were reviewed and corrected by a board-certified radiologist and  ...
```
Why are nine cases mentioned here but `summary.data.evaluation.vol_samples = 7`?

vanossj replied:
7 is correct

LennyN95 commented on models/bamf_nnunet_mr_liver/meta.json:
```diff
@@ -0,0 +1,137 @@
+        },
+        "evaluation": {
+            "title": "Evaluation Data",
+            "text": "The model was used to segment cases from the IDC [1] collection TCGA-LIHC [1]. Nine of those cases were randomly selected to be reviewed and corrected by a board-certified radiologist. The model predictions, and radiologist corrections are published on zenodo [3]",
```
Why are nine cases mentioned here but `summary.data.evaluation.vol_samples = 7`?

vanossj committed:
correct the evalutation data size

LennyN95 commented:
@rahul99 could you upload 1-5 images for the model card that show the model in action? This is not a hard requirement, thus I moved on with the push. However, from a UX perspective, visuals add a great value to the model page and also help readers to quickly recognize a model (especially as the model card is made to look similar across all models, images can be a great help here).
